crates/metaslang/cst/src/text_index.rs

```diff
@@ -1,22 +1,46 @@
 use std::fmt::Display;
-use std::ops::{Add, AddAssign, Range, Sub, SubAssign};
+use std::ops::{Add, AddAssign, Range};
 
 use serde::Serialize;
 
-#[derive(Default, Copy, Clone, PartialEq, Eq, Debug, Serialize)]
+#[derive(Default, Hash, Copy, Clone, PartialEq, Eq, Debug, Serialize)]
 pub struct TextIndex {
     pub utf8: usize,
     pub utf16: usize,
-    pub char: usize,
+    pub line: usize,
+    pub column: usize,
 }
 
 impl TextIndex {
-    /// Shorthand for `TextIndex { utf8: 0, utf16: 0, char: 0 }`.
+    /// Shorthand for `TextIndex { utf8: 0, utf16: 0, line: 0, char: 0 }`.
     pub const ZERO: TextIndex = TextIndex {
         utf8: 0,
         utf16: 0,
-        char: 0,
+        line: 0,
+        column: 0,
     };
+
+    #[inline]
+    pub fn advance(&mut self, c: char, next: Option<&char>) {
+        self.utf8 += c.len_utf8();
+        self.utf16 += c.len_utf16();
+        if c == '\n'
```

Comment:
style nitpick: This could probably be rewritten for more clarity as
```rust
match (c, next) {
  ('\n', _) => ...,
  ('\r', next) if next != Some(&'\n') => ...
  _ => ...
```

Reply:
I note, ironically, that the code you suggest hides an error - `'\r'` followed by `'\n'` shouldn't advance the column. So even thouogh I prefer `match` and your solution, I'm not sure that it is more clear.

Reply:
My bad, this was written quickly. Something like this, then:

```rust
match (c, next) {
  ('\n' | '\u{2028}' | '\u{2029}, _) => ...,
  ('\r', Some('\n')) => {/* Ignore for now, we will already advance the column once we skip to \n, per above */},
  ('\r', _) => ...
  _ => ...
  ```
  
  which should be more clear; the nested condition in the `\r` was subtle enough that it I didn't completely get it while glancing at the code, whereas this should be more clear with the comment etc.

Reply:
Also see https://github.com/NomicFoundation/slang/pull/918#discussion_r1604939685; there's a chance that simply accounting for `\n` is the pragmatic choice here.